File: low-level-design/design-patterns/python/behavioural/state-pattern/starting_state.py
from state import VMState




class StartingError(Exception):
    pass 


class StartingState(VMState):
    def start(self, vm):
        print(f"VM {vm.name} is already in started state, no action performed")
        return StartingError("VM is already in started state, no action performed")
    
    def stop(self, vm):
        from state_factory import StateFactory
        print(f"Stopping VM {vm.name}")
        vm.set_state(StateFactory.get_stopped_state())
    
    def restart(self, vm):
        from state_factory import StateFactory
        print(f"Restarting VM {vm.name}")
        vm.set_state(StateFactory.get_restarting_state())
    
    def suspend(self, vm):
        from state_factory import StateFactory
        print(f"Suspending VM {vm.name}")
        vm.set_state(StateFactory.get_suspended_state())
    
    def get_name(self):
        return "Starting"

File: low-level-design/design-patterns/golang/behavioural/state-pattern/state/starting_state.go
package state

import (
	"errors"
	"fmt"
)

func NewStartingState() *StartingState {
	return &StartingState{}
}

type StartingState struct {
}

func (s *StartingState) Start(vm *VM) error {
	fmt.Printf("%s vm is already starting \n",vm.name)
	return errors.New("vm is already starting")
}

func (s *StartingState) Stop(vm *VM) error {
	fmt.Printf("%s stopping the vm \n",vm.name)
	vm.SetState(NewStoppedState())
	return nil
}

func (s *StartingState) Restart(vm *VM) error {
	fmt.Printf("%s restarting the vm \n",vm.name)
	vm.SetState(NewRestartingState())
	return nil
}

func (s *StartingState) Suspend(vm *VM) error {
	fmt.Printf("%s suspending the vm \n",vm.name)
	vm.SetState(NewSuspendedState())
	return nil
}
func (s *StartingState) GetName() string {
	return "Starting"
}
